### PR TITLE
Can query ldap groups can now be queried for a user to perform group …

### LINKED ISSUE #807: fetch ldap groups for authenticated users
to be able to do group based authorization to products

### PR DESCRIPTION
**Pull request by dkrupp**
…based access control.

A bug was fixed:the authentication token cookie was not always set in the response header
which caused frequent re-authnentications.

Closes #807.

### CODE REVIEW

File: docs/authentication.md [hunk 1]
```diff
@@ -191,15 +191,13 @@ servers as it can elongate the authentication process.
...
-  queried user account DN.
-
-  Example configuration: `(member=$USERDN$)`
+   Group query pattern used LDAP query expression to find the group objects
+   a user is a member of. It must contain a `$USERDN$` pattern. 
+   `$USERDN$` will be automatically replaced by the queried user account DN.
+   
+ * `groupNameAttr`
+ 	
+ 	The attribute of the group object which contains the name of the group.
```
Comment (whisperity) [inline docs/authentication.md]: Keep the indentation in this line.

File: libcodechecker/libauth/cc_ldap.py [hunk 2]
```diff
@@ -307,17 +341,36 @@ def auth_user(ldap_config, username=None, credentials=None):
...
+                      'Its value must be the name'
+                      'attribute of the group.')
+            return []
 
-        member_pattern = member_pattern.replace('$USERDN$', user_dn)
-        member_query = '(& ' + group_pattern + member_pattern + ')'
+        # Attribute name must be ascii encoded
+        group_name_attr = group_name_attr.encode('ascii',
+                                                 'ignore')
+        attr_list = []
```
Comment (whisperity) [inline libcodechecker/libauth/cc_ldap.py]: This list and the `append()` call can be simplified, as you only append once to the list:

```py
attr_list = [group_name_attr]
```

File: libcodechecker/libauth/cc_ldap.py [hunk 3]
```diff
@@ -307,17 +341,36 @@ def auth_user(ldap_config, username=None, credentials=None):
...
+        LOG.debug("Performing LDAP search for group:" + group_pattern +
+                  "Group Name Attr:" + group_name_attr)
+
+        with ldap_error_handler():
+            group_result = connection.search_s(group_base,
+                                               group_scope,
+                                               group_pattern,
+                                               attr_list)
+        groups = []
+        for g in group_result:
```
Comment (gyorb) [inline libcodechecker/libauth/cc_ldap.py]: `group_result` will be undefined if there is some error in the `search_s`

File: libcodechecker/libauth/cc_ldap.py [hunk 4]
```diff
@@ -307,17 +341,36 @@ def auth_user(ldap_config, username=None, credentials=None):
...
+            groups.append(g[1][group_name_attr][0])
```
Comment (gyorb) [inline libcodechecker/libauth/cc_ldap.py]: Could you add some example what is the expected group format which is processed here?

File: libcodechecker/server/client_db_access_server.py [hunk 5]
```diff
@@ -127,6 +128,17 @@ def __check_auth_in_request(self):
 
         return success
 
+    def end_headers(self):
+        # Sending the authentication cookie
+        # in every response if any.
```
Comment (gyorb) [inline libcodechecker/server/client_db_access_server.py]: I would add some extra comment here too, why this is needed.

**Comment by bruntib** [comment]
Python dependencies for LDAP are listed in `.ci/auth_requirements`. No makefile installs these dependencies and no documentation mentions that it is necessary.

**Pull request merged**